apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/DatasetsResourceTest.java

```diff
@@ -3135,9 +3137,74 @@ void getDatasetItemsByDatasetId__whenItemsWereUpdated__thenReturnCorrectItemsCou
             }
         }
 
+        @Test
+        @DisplayName("when items have data with same keys and different types, then return columns types and count")
+        void getDatasetItemsByDatasetId__whenItemsHaveDataWithSameKeysAndDifferentTypes__thenReturnColumnsTypesAndCount() {
+
+            UUID datasetId = createAndAssert(factory.manufacturePojo(Dataset.class).toBuilder()
+                    .id(null)
+                    .build());
+
+            var item = factory.manufacturePojo(DatasetItem.class);
+
+            var item2 = item.toBuilder()
+                    .id(factory.manufacturePojo(UUID.class))
+                    .data(item.data()
+                            .keySet()
+                            .stream()
+                            .map(key -> Map.entry(key, NullNode.getInstance()))
+                            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue)))
+                    .build();
+
+            var item3 = item.toBuilder()
+                    .id(factory.manufacturePojo(UUID.class))
+                    .data(item.data()
+                            .keySet()
+                            .stream()
+                            .map(key -> Map.entry(key, TextNode.valueOf(RandomStringUtils.randomAlphanumeric(10))))
+                            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue)))
+                    .build();
+
+
+            var batch = factory.manufacturePojo(DatasetItemBatch.class).toBuilder()
+                    .items(List.of(item, item2, item3))
+                    .datasetId(datasetId)
+                    .build();
+
+            List<Map<String, JsonNode>> data = batch.items()
+                    .stream()
+                    .map(DatasetItem::data)
+                    .toList();
+
+            Set<Column> columns = addDeprecatedFields(data);
+
+            putAndAssert(batch, TEST_WORKSPACE, API_KEY);
+
+            try (var actualResponse = client.target(BASE_RESOURCE_URI.formatted(baseURI))
+                    .path(datasetId.toString())
+                    .path("items")
+                    .request()
+                    .header(HttpHeaders.AUTHORIZATION, API_KEY)
+                    .header(WORKSPACE_HEADER, TEST_WORKSPACE)
+                    .get()) {
+
+                assertThat(actualResponse.getStatusInfo().getStatusCode()).isEqualTo(200);
+
+                var actualEntity = actualResponse.readEntity(DatasetItemPage.class);
+
+                assertThat(actualEntity.size()).isEqualTo(batch.items().size());
+                assertThat(actualEntity.content()).hasSize(batch.items().size());
+                assertThat(actualEntity.page()).isEqualTo(1);
+                assertThat(actualEntity.total()).isEqualTo(batch.items().size());
+                assertThat(actualEntity.columns()).isEqualTo(columns);
+
+                assertPage(batch.items().reversed(), actualEntity.content());
+            }
```

Comment:
Minor: can you double check that not copied and pasted? If so, reuse whatever aux method.

Reply:
It's not. If you check how the items are created, it's different to guarantee different types. But I can try to extract the assertion of the page.